kotlin-analysis-core/src/main/kotlin/org/jetbrains/research/ml/kotlinAnalysis/Analyzers.kt

```diff
@@ -0,0 +1,119 @@
+package org.jetbrains.research.ml.kotlinAnalysis
+
+import com.intellij.psi.PsiElement
+import com.intellij.psi.PsiRecursiveElementVisitor
+
+typealias AnalyzerToStat<P, R> = Map<PsiSecondaryAnalyzer<P, R>, MutableMap<P, R>>
+
+/**
+ * Classes that inherit from this interface implement secondary analyzers that analyze the concrete type of PSI elements.
+ *
+ * @param P the type of PSI elements, that can be analyzed.
+ * @param R the type of the analysis result.
+ */
+interface PsiSecondaryAnalyzer<P : PsiElement, R> {
+    /**
+     * Returns the analysis result of a [PSI element][psiElement].
+     */
+    fun analyze(psiElement: P): R
+}
+
+/**
+ * Classes that inherit from this interface implement secondary analyzers with a cache.
+ * These analyzers analyze the concrete type of PSI elements. The cache is used to memorize the results of the analysis.
```

Comment:
Please, add an example of usage:
```
Classes that inherit from this interface implement analyzer that analyzes the concrete type of PSI elements with a cache. The cache is used to memorize the results of the analysis.

For example, for a task "check if a class has only <var> fields" we can cache the result of this analysis if the class has several fields with the same type to avoid analysing the same objects twice. 
```